File: 1443_minimum_time_to_collect_all_apples_in_a_tree/solution.js
var minTime = function(n, edges, hasApple) {
    const graph = Array.from({length: n}, () => []); for (const [a, b] of edges) { graph[a].push(b); graph[b].push(a); }
    const dfs = (node, parent) => { let time = 0; for (const child of graph[node]) if (child !== parent) { const childTime = dfs(child, node); if (childTime || hasApple[child]) time += childTime + 2; } return time; };
    return dfs(0, -1);
};

File: 1443_minimum_time_to_collect_all_apples_in_a_tree/solution.rb
# LeetCode 1443 - Minimum Time To Collect All Apples In A Tree
# https://leetcode.com/problems/minimum-time-to-collect-all-apples-in-a-tree/

def min_time(n, edges, has_apple)
  graph = Array.new(n) { [] }
  edges.each do |a, b|
    graph[a] << b
    graph[b] << a
  end
  visit = lambda do |node, parent|
    cost = 0
    graph[node].each do |child|
      next if child == parent
      child_cost = visit.call(child, node)
      cost += child_cost + 2 if child_cost > 0 || has_apple[child]
    end
    cost
  end
  visit.call(0, -1)
end
